stlab/forest.hpp

```diff
@@ -0,0 +1,1240 @@
+    }
+
+    bool equal(const forest_const_iterator& y) const {
+        return (_node == y._node) && (_edge == y._edge);
+    }
+
+    node_t* _node{nullptr};
+    forest_edge _edge{forest_edge::leading};
+
+    forest_const_iterator(node_t* node, forest_edge edge) : _node(node), _edge(edge) {}
+};
+
+/**************************************************************************************************/
+
+} // namespace detail
+
+/**************************************************************************************************/
+
+namespace unsafe {
+
+/**************************************************************************************************/
+
+template <typename T> // T is node<T>
+struct set_next_fn<detail::forest_iterator<T>> {
+    void operator()(detail::forest_iterator<T> x, detail::forest_iterator<T> y) const {
+        using node_t = typename detail::node<T>;
+
+        x._node->link(x.edge(), node_t::next_s) = y._node;
+        y._node->link(y.edge(), node_t::prior_s) = x._node;
+    }
+};
+
+/**************************************************************************************************/
+
+} // namespace unsafe
+
+/**************************************************************************************************/
+
+template <typename T>
+class forest {
+private:
+    using node_t = detail::node<T>;
+    friend class child_adaptor<forest<T>>;
+
+public:
+    // types
+    using reference = T&;
+    using const_reference = const T&;
+    using iterator = detail::forest_iterator<T>;
+    using const_iterator = detail::forest_const_iterator<T>;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using value_type = T;
+    using pointer = T*;
+    using const_pointer = const T*;
+    using reverse_iterator = reverse_fullorder_iterator<iterator>;
+    using const_reverse_iterator = reverse_fullorder_iterator<const_iterator>;
+
+    using child_iterator = child_iterator<iterator>;
+    /* qualification needed since: A name N used in a class S shall refer to the same declaration
+       in its context and when re-evaluated in the completed scope of
+       S. */
+    using const_child_iterator = stlab::child_iterator<const_iterator>;
+    using reverse_child_iterator = std::reverse_iterator<child_iterator>;
+
+    using preorder_iterator = edge_iterator<iterator, forest_edge::leading>;
+    using const_preorder_iterator = edge_iterator<const_iterator, forest_edge::leading>;
+    using postorder_iterator = edge_iterator<iterator, forest_edge::trailing>;
+    using const_postorder_iterator = edge_iterator<const_iterator, forest_edge::trailing>;
+
+    forest() = default;
+    ~forest() { clear(); }
+
+    forest(const forest&);
+    forest(forest&&) noexcept;
+    forest& operator=(const forest& x) {
+        auto tmp = x;
+        *this = std::move(tmp);
+        return *this;
+    }
+    forest& operator=(forest&& x) noexcept {
+        clear();
+        splice(end(), x);
+        return *this;
```

Comment:
I've been having a long debate about the postconditions of self-move-assignment with Herb and Howard. The standard library is currently all over the place (with some implementations crashing on some types). std::unique_ptr<> actually defines it to be a noop and Herb would like all the standard types to guarantee this. The argument basically boils down to a more general rule that to avoid issues with reference aliasing, you should read everything that could alias first, then write. For `std::unique_ptr` move assignment is defined as `reset(r.release())`. That has me thinking about idiomatic move-assignment. How about this:

```suggestion
        auto tmp{move(x)}; // this is `release()`
        clear(); // these two lines are `reset()`
        splice(end(), tmp);
        return *this;
```

It is mostly unnecessary extra work (which I dislike) but may be the direction the standard is moving. I've started what is going to be a defect report around move (and it has expanded to a larger issue of defining the domain of operations) - https://sean-parent.stlab.cc/draft/2020/02/26/revisionist-history-of-move.html.

Another option is to collapse both assignments with pass-by-value

```
    forest& operator=(forest x) noexcept {
        clear();
        splice(end(), x);
        return *this;
```
Initially in C++11 that didn't work, but it was fixed, retroactively, with a defect report.